toolchain/check/type.cpp

```diff
@@ -163,4 +163,33 @@ auto GetUnboundElementType(Context& context, SemIR::TypeInstId class_type_id,
                                                 element_type_id);
 }
 
+auto GetCanonicalizedFacetOrTypeValue(Context& context, SemIR::InstId inst_id)
+    -> SemIR::InstId {
+  // We can have FacetAccessType of a FacetValue, and a FacetValue of a
+  // FacetAccessType, but they don't nest indefinitely.
+  if (auto access = context.insts().TryGetAs<SemIR::FacetAccessType>(inst_id)) {
+    inst_id = access->facet_value_inst_id;
+  }
+  if (auto value = context.insts().TryGetAs<SemIR::FacetValue>(inst_id)) {
+    inst_id = value->type_inst_id;
+
+    if (auto access =
+            context.insts().TryGetAs<SemIR::FacetAccessType>(inst_id)) {
+      inst_id = access->facet_value_inst_id;
+    }
+  }
+
+  CARBON_DCHECK(!context.insts().Is<SemIR::FacetAccessType>(inst_id));
+  CARBON_DCHECK(!context.insts().Is<SemIR::FacetValue>(inst_id));
```

Comment:
Usually we use CARBON_CHECK instead of DCHECK unless we have a specific performance concern.
```suggestion
  CARBON_CHECK(!context.insts().Is<SemIR::FacetAccessType>(inst_id));
  CARBON_CHECK(!context.insts().Is<SemIR::FacetValue>(inst_id));
```